cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,316 @@
+   */
+   long getMatchingOffset(final Checkpoint checkpoint, final int partition) {
+    // Get BrokerInfo for constructing SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      return kafka.api.OffsetRequest.EarliestTime(); // Return earliest offset if brokerInfo is null
+    }
+    SimpleConsumer simpleConsumer
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT_MILLIS, FETCH_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    long timeStamp = getEventTimeByOffset(simpleConsumer, partition, clientId, checkpoint.getNextOffset() - 1);
+    if (timeStamp == checkpoint.getNextEventTime()) {
+      return checkpoint.getNextOffset();
+    }
+
+    long smallestMatchingOffset = findOffsetByTime(simpleConsumer, partition, clientId, checkpoint.getNextEventTime());
+    if (smallestMatchingOffset >= 0) {
+      // Increment smallestMatchingOffset to get the next offset if it's a valid non-negative offset
+      smallestMatchingOffset++;
+    }
+    return smallestMatchingOffset;
+  }
+
+  /**
+   * Determine the lower-bound and upper-bound of offsets to search for the smallest offset of the message
+   * with log event time equal to {@code targetTime}, and then perform the search.
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long findOffsetByTime(SimpleConsumer consumer, int partition, String clientName, long targetTime) {
+    TopicAndPartition topicAndPartition = new TopicAndPartition(topic, partition);
+    Map<TopicAndPartition, PartitionOffsetRequestInfo> requestInfo = new HashMap<>();
+    // the offset of the earliest message
+    long minOffset = fetchOffsetByTime(consumer, partition, kafka.api.OffsetRequest.EarliestTime());
+    // the next offset after the latest message
+    long maxOffset = fetchOffsetByTime(consumer, partition, kafka.api.OffsetRequest.LatestTime());
+    long endOffset = maxOffset - 1; // the offset of the latest message
+    long prevOffset = Long.MAX_VALUE;
+    // Start from (targetTime + replicationDelayMillis) to be safer
+    long requestTime = targetTime + replicationDelayMillis;
+    while (true) {
+      requestInfo.put(topicAndPartition, new PartitionOffsetRequestInfo(requestTime, 1));
+      OffsetResponse response =
+        consumer.getOffsetsBefore(new OffsetRequest(requestInfo, kafka.api.OffsetRequest.CurrentVersion(), clientName));
+      // Fetch the starting offset of the last segment whose latest latest message is published before requestTime
+      long[] responseOffsets = response.offsets(topic, partition);
```

Comment:
No error handling? Why not refactor the `fetchOffsetByTime` method to make it works for this case?